README.md
# Arranged.StackView

<p align="left">
<img src="https://img.shields.io/cocoapods/v/Arranged.svg?label=version">
<a href="https://github.com/Carthage/Carthage"><img src="https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat"></a>
<a href="http://cocoadocs.org/docsets/Arranged"><img src="https://img.shields.io/cocoapods/p/Arranged.svg?style=flat)"></a>
</p>

Open source replacement of [UIStackView](https://developer.apple.com/library/prerelease/ios/documentation/UIKit/Reference/UIStackView_Class_Reference/) for iOS 8 (100% layouts supported)

> **Deprecated**. UIStackView was introduces in iOS 9, please use the native one.

<img src="https://cloud.githubusercontent.com/assets/1567433/13573981/364b2946-e493-11e5-9d02-893a5dc11a8c.png" width="50%"/>

- Supports all `alignments` and `distributions`, `spacing`, `baselineRelativeArrangement`, `layoutMarginsRelativeArrangement`, `axis`
- Unit tested, thousands of layouts covered
- Supports animations
- Generates exactly the same sets of constraints as `UIStackView`:

**UIStackView** (`Alignment.Leading`, `Distribution.FillEqually`)
```
<'UISV-alignment' content-view-1.top == content-view-2.top>
<'UISV-canvas-connection' UIStackView:0x7f9cf4816930.leading == content-view-1.leading>
<'UISV-canvas-connection' H:[content-view-2]-(0)-|>
<'UISV-canvas-connection' UIStackView:0x7f9cf4816930.top == content-view-1.top>
<'UISV-canvas-connection' V:[_UILayoutSpacer:0x7f9cf4849f80'UISV-alignment-spanner']-(0)-|>
<'UISV-fill-equally' content-view-2.width == content-view-1.width>
<'UISV-spacing' H:[content-view-1]-(10)-[content-view-2]>
<'UISV-spanning-boundary' _UILayoutSpacer:0x7f9cf4849f80'UISV-alignment-spanner'.top == content-view-1.top priority:999.5>
<'UISV-spanning-boundary' _UILayoutSpacer:0x7f9cf4849f80'UISV-alignment-spanner'.bottom >= content-view-1.bottom>
<'UISV-spanning-boundary' _UILayoutSpacer:0x7f9cf4849f80'UISV-alignment-spanner'.top == content-view-2.top priority:999.5>
<'UISV-spanning-boundary' _UILayoutSpacer:0x7f9cf4849f80'UISV-alignment-spanner'.bottom >= content-view-2.bottom>
<'UISV-spanning-fit' V:[_UILayoutSpacer:0x7f9cf4849f80'UISV-alignment-spanner'(0@51)] priority:51>
<'UISV-ambiguity-suppression' V:[content-view-1(0@25)] priority:25>
<'UISV-ambiguity-suppression' V:[content-view-2(0@25)] priority:25>
```

**Arranged** (`Alignment.Leading`, `Distribution.FillEqually`)
```
<'ASV-alignment' content-view-1.top == content-view-2.top>
<'ASV-canvas-connection' Arranged.StackView:0x7f9cf4822c90.leading == content-view-1.leading>
<'ASV-canvas-connection' H:[content-view-2]-(0)-|>
<'ASV-canvas-connection' Arranged.StackView:0x7f9cf4822c90.top == content-view-1.top>
<'ASV-canvas-connection' V:[Arranged.LayoutSpacer:0x7f9cf2c4f3f0]-(0)-|>
<'ASV-fill-equally' content-view-1.width == content-view-2.width>
<'ASV-spacing' H:[content-view-1]-(10)-[content-view-2]>
<'ASV-spanning-boundary' Arranged.LayoutSpacer:0x7f9cf2c4f3f0.top == content-view-1.top priority:999.5>
<'ASV-spanning-boundary' Arranged.LayoutSpacer:0x7f9cf2c4f3f0.bottom >= content-view-1.bottom>
<'ASV-spanning-boundary' Arranged.LayoutSpacer:0x7f9cf2c4f3f0.top == content-view-2.top priority:999.5>
<'ASV-spanning-boundary' Arranged.LayoutSpacer:0x7f9cf2c4f3f0.bottom >= content-view-2.bottom>
<'ASV-spanning-fit' V:[Arranged.LayoutSpacer:0x7f9cf2c4f3f0(0@51)] priority:51>
<'ASV-ambiguity-suppression' V:[content-view-1(0@25)] priority:25>
<'ASV-ambiguity-suppression' V:[content-view-2(0@25)] priority:25>
```

## Usage

`Arranged.StackView` is used the same way `UIStackView` is:

```swift
let stackView = StackView(arrangedSubviews: [view1, view2, view3])
stackView.alignment = .leading
stackView.distribution = .fillEqually
stackView.spacing = 20
stackView.axis = .vertical
stackView.isLayoutMarginsRelativeArrangement = true
```

The only difference is hiding items:

```swift
UIView.animateWithDuration(0.33) {
    stackView.setArrangedView(view, hidden: true)
    stackView.layoutIfNeeded()
}
```

## Requirements

- iOS 8.0, tvOS 9.0
- Xcode 9
- Swift 4

## Getting Started

- Get a demo project using `pod try Arranged` command
- [Install](#installation), `import Arranged` and enjoy!

## Differences

- `UIStackView` observes `hidden` property of arranged views, delays its effect if called inside animation block, and updates constraints accordingly. I believe this behavior is confusing and impractical to implement. `Arranged.StackView` provides a straightforward method `setArrangedView(_:hidden:)` which updates constraints exactly the same way as `UIStackView` does, but it doesn't affect `hidden` property.
- Animations require you to call `view.layoutIfNeeded()` - just like with any regular layout
- `StackViewDistribution.FillProportionally` doesn't update its constrains when `intrinsicContentSize` of arranged views changes due to the fact that `UIStackView` uses private API (`_intrinsicContentSizeInvalidatedForChildView`) to achieve that
- `UISV-text-width-disambiguation` constraints are not implemented because they are not documented (and you probably should disambiguate text views the way that fits your app anyway)

## Installation<a name="installation"></a>

### [CocoaPods](http://cocoapods.org)

To install Arranged add a dependency to your Podfile:

```ruby
# source 'https://github.com/CocoaPods/Specs.git'
# use_frameworks!
# platform :ios, "8.0"

pod "Arranged"
```

### [Carthage](https://github.com/Carthage/Carthage)

To install Arranged add a dependency to your Cartfile:

```
github "kean/Arranged"
```

### Import

Import installed modules in your source files

```swift
import Arranged
```

## License

Arranged is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
[Changelog](https://github.com/kean/Arranged/releases) for all versions

## Arranged 2.2

- Make UIStackView APIs open

## Arranged 2.1

- Updated to Swift 4

## Arranged 2.0
 
- Updated to Swift 3

## Arranged 1.0.4

- Activate constraints using `activate` property instead of manually adding them to the view
- `StackView` now returns true in `requiresConstraintBasedLayout`
- `LayoutGuides` are now hidden by default
- `LayoutGuides` now have `translatesAutoresizingMaskIntoConstraints` set to false be default

## Arranged 1.0.3
 
- `insertArrangedSubview(_:atIndex:)` method now updates the stack index (but not the subview index) of the arranged subview if it's already in the `arrangedSubviews` list.
- Add app extensions support
- Add tvOS support (useful if you share code between iOS 8.0 and tvOS)

## Arranged 1.0.2

- Make _viewForFirstBaselineLayout private (was made public by mistake)
- Update documentation

## Arranged 1.0.1

- More efficient constraint deactivation on updateConstraints()
- Cleaner arrangements implementation

## Arranged 1.0

- Initial release


Tests/Adapter.swift
// The MIT License (MIT)
//
// Copyright (c) 2018 Alexander Grebenyuk (github.com/kean).

import Foundation
import UIKit
import Arranged

protocol StackViewAdapter: class {
    // Those members are the same for both classes
    func addArrangedSubview(_ view: UIView)
    func removeArrangedSubview(_ view: UIView)
    var axis: UILayoutConstraintAxis { get set }
    var spacing: CGFloat { get set }
    var isBaselineRelativeArrangement: Bool { get set }
    var isLayoutMarginsRelativeArrangement: Bool { get set }
    
    var ar_distribution: UIStackViewDistribution { get set }
    var ar_alignment: UIStackViewAlignment { get set }
    
    func setArrangedView(_ view: UIView, hidden: Bool)
}

extension StackView: StackViewAdapter {
    var ar_distribution: UIStackViewDistribution {
        get { return self.distribution.toStackViewDistrubition() }
        set { self.distribution = StackViewDistribution.fromStackViewDistrubition(newValue) }
    }
    var ar_alignment: UIStackViewAlignment {
        get { return self.alignment.toStackViewAlignment() }
        set { self.alignment = StackViewAlignment.fromStackViewAlignment(newValue) }
    }
}

extension UIStackView: StackViewAdapter {
    var ar_distribution: UIStackViewDistribution {
        get { return self.distribution }
        set { self.distribution = newValue }
    }
    var ar_alignment: UIStackViewAlignment {
        get { return self.alignment }
        set { self.alignment = newValue }
    }
    func setArrangedView(_ view: UIView, hidden: Bool) {
        view.isHidden = hidden
    }
}

extension StackViewDistribution {
    func toStackViewDistrubition() -> UIStackViewDistribution {
        switch self {
        case .fill: return .fill
        case .fillEqually: return .fillEqually
        case .fillProportionally: return .fillProportionally
        case .equalSpacing: return .equalSpacing
        case .equalCentering: return .equalCentering
        }
    }
    
    static func fromStackViewDistrubition(_ distribution: UIStackViewDistribution) -> StackViewDistribution {
        switch distribution {
        case .fill: return .fill
        case .fillEqually: return .fillEqually
        case .fillProportionally: return .fillProportionally
        case .equalSpacing: return .equalSpacing
        case .equalCentering: return .equalCentering
        }
    }
}

extension StackViewAlignment {
    func toStackViewAlignment() -> UIStackViewAlignment {
        switch self {
        case .fill: return .fill
        case .leading: return .leading
        case .firstBaseline: return .firstBaseline
        case .center: return .center
        case .trailing: return .trailing
        case .lastBaseline: return .lastBaseline
        }
    }
    
    static func fromStackViewAlignment(_ alignment: UIStackViewAlignment) -> StackViewAlignment {
        switch alignment {
        case .fill: return .fill
        case .leading: return .leading
        case .firstBaseline: return .firstBaseline
        case .center: return .center
        case .trailing: return .trailing
        case .lastBaseline: return .lastBaseline
        }
    }
}


Tests/Internal.swift
// The MIT License (MIT)
//
// Copyright (c) 2018 Alexander Grebenyuk (github.com/kean).

import Arranged
import UIKit
import XCTest

func assertEqualConstraints(_ constraints1: [NSLayoutConstraint], _ constraints2: [NSLayoutConstraint]) -> Bool {
    func filterOutMarginContraints(_ constraints: [NSLayoutConstraint]) -> [NSLayoutConstraint] {
        return constraints.filter {
            return $0.identifier?.hasSuffix("Margin-guide-constraint") == false
        }
    }
    // We fitler out margin-guide constraints because Arranged.StackView doesn't use those
    return _assertEqualConstraints(filterOutMarginContraints(constraints1), filterOutMarginContraints(constraints2))
}

func _assertEqualConstraints(_ constraints1: [NSLayoutConstraint], _ constraints2: [NSLayoutConstraint]) -> Bool {
    guard constraints1.count == constraints2.count else {
        XCTFail("Constraints count doesn't match")
        return false
    }
    
    var array1 = constraints1
    var array2 = constraints2
    
    while let c1 = array1.popLast() {
        let idx2 = array2.index { c2 in
            return isEqual(c1, c2)
        }
        guard let unpackedIdx2 = idx2 else {
            XCTFail("Couldn't find matching constraint for \(c1)")
            return false
        }
        array2.remove(at: unpackedIdx2)
    }
    
    guard array1.count == 0 && array2.count == 0 else {
        XCTFail("Failed to match all constraints")
        return false
    }
    
    return true
}

// MARK: Constrain Comparison

func isEqual(_ lhs: NSLayoutConstraint, _ rhs: NSLayoutConstraint) -> Bool {
    func identifier(_ constraint: NSLayoutConstraint) -> String? {
        return constraint.identifier?.replacingOccurrences(of: "ASV-", with: "UISV-")
    }
    
    guard identifier(lhs) == identifier(rhs) else {
        return false
    }
        
    func isEqual(_ item1: AnyObject?, _ item2: AnyObject?) -> Bool {
        // True if both nil
        if item1 == nil && item2 == nil {
            return true
        }
        // True if both items are stack views
        if ((item1 is UIStackView && item2 is Arranged.StackView) ||
            (item1 is Arranged.StackView && item2 is UIStackView)) {
            return true
        }
        // True if both are for content views with the same indexes
        if let view1 = item1 as? UIView, let view2 = item2 as? UIView {
            return view1.test_isContentView && view2.test_isContentView && view1.tag == view2.tag
        }
        // FIXME: Find a better way to test layout guides
        // Assume that the remaining items are spacers, we can't check really
        return item1 != nil && item2 != nil
    }
    guard isEqual(lhs.firstItem, rhs.firstItem) &&
        isEqual(lhs.secondItem, rhs.secondItem) else {
            return false
    }    
    return lhs.firstAttribute == rhs.firstAttribute &&
        lhs.secondAttribute == rhs.secondAttribute &&
        lhs.relation == rhs.relation &&
        lhs.constant == rhs.constant &&
        lhs.multiplier == rhs.multiplier &&
        lhs.priority == rhs.priority
}


// MARK: Misc

func constraintsFor(_ view: UIView) -> [NSLayoutConstraint] {
    var constraints = [NSLayoutConstraint]()
    constraints.append(contentsOf: view.constraints)
    for subview in view.subviews {
        constraints.append(contentsOf: subview.constraints)
    }
    return constraints
}

extension UIStackViewAlignment {
    var toString: String {
        switch self {
        case .fill: return ".Fill"
        case .leading: return ".Leading"
        case .firstBaseline: return ".FirstBaseline"
        case .center: return ".Center"
        case .lastBaseline: return ".LastBaseline"
        case .trailing: return ".Trailing"
        }
    }
}

extension UIStackViewDistribution {
    var toString: String {
        switch self {
        case .fill: return ".Fill"
        case .fillEqually: return ".FillEqually"
        case .fillProportionally: return ".FillProportionally"
        case .equalSpacing: return ".EqualSpacing"
        case .equalCentering: return ".EqualCentering"
        }
    }
}


private struct AssociatedKeys {
    static var IsContentView = "Arranged.Test.IsContentView"
}

extension UIView {
    var test_isContentView: Bool {
        get {
            return objc_getAssociatedObject(self, &AssociatedKeys.IsContentView) != nil
        }
        set {
            objc_setAssociatedObject(self, &AssociatedKeys.IsContentView, true, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        }
    }
}


Tests/LayoutTests.swift
// The MIT License (MIT)
//
// Copyright (c) 2018 Alexander Grebenyuk (github.com/kean).

import Arranged
import UIKit
import XCTest

var testCasesCount = 0
var failedTestCasesCount = 0

var constraintsPrinted = 0
let maxConstraintsPrinted = 50


class LayoutTests: XCTestCase {
    func testWithoutItems() {
        printTestTitle("Test: 0 views")
        _test{ return [] }
    }
    
    func testWithItems() {
        printTestTitle("Test: 1 content view")
        _test{ return [ContentView()] }
        
        printTestTitle("Test: 3 content view")
        _test{ return [ContentView(), ContentView(), ContentView()] }
    }
    
    func testWithItemSomeOfWhichHaveNoIntrinsic() {
        printTestTitle("Test: 3 content views, 1 without intrinsic")
        _test{
            return [ContentView(),
                ContentView(contentSize: CGSize(width: UIViewNoIntrinsicMetric, height: UIViewNoIntrinsicMetric)),
                ContentView()]
        }
        
        printTestTitle("Test: 3 content views, 2 without intrinsic")
        _test {
            return [ContentView(contentSize: CGSize(width: UIViewNoIntrinsicMetric, height: UIViewNoIntrinsicMetric)),
                ContentView(),
                ContentView(contentSize: CGSize(width: UIViewNoIntrinsicMetric, height: UIViewNoIntrinsicMetric))]
        }
        
        printTestTitle("Test: 3 content views, 3 without intrinsic")
        _test {
            return [ContentView(contentSize: CGSize(width: UIViewNoIntrinsicMetric, height: UIViewNoIntrinsicMetric)),
            ContentView(contentSize: CGSize(width: UIViewNoIntrinsicMetric, height: UIViewNoIntrinsicMetric)),
            ContentView(contentSize: CGSize(width: UIViewNoIntrinsicMetric, height: UIViewNoIntrinsicMetric))]
        }
    }
    
    func testWithHiddenItems() {
        printTestTitle("Test: 1 content view, hidden")
        _test(views: {
            return [ContentView()]
        }, update: { stack, views in
            stack.setArrangedView(views[0], hidden: true)
        })
        
        printTestTitle("Test: 2 content views, 1st hidden")
        _test(views: {
            return [ContentView(), ContentView()]
        }, update: { stack, views in
            stack.setArrangedView(views[0], hidden: true)
        })
        
        printTestTitle("Test: 2 content views, 2nd hidden")
        _test(views: {
            return [ContentView(), ContentView()]
        }, update: { stack, views in
            stack.setArrangedView(views[1], hidden: true)
        })
        
        printTestTitle("Test: 2 content views, both hidden")
        _test(views: {
            return [ContentView(), ContentView()]
        }, update: { stack, views in
            stack.setArrangedView(views[0], hidden: true)
            stack.setArrangedView(views[1], hidden: true)
        })
        
        printTestTitle("Test: 3 content views, all hidden")
        _test(views: {
            return [ContentView(), ContentView(), ContentView()]
        }, update: { stack, views in
            stack.setArrangedView(views[0], hidden: true)
            stack.setArrangedView(views[1], hidden: true)
            stack.setArrangedView(views[2], hidden: true)
        })
        
        /* This test is disabled because UIStackView constructs invalid spacing constraint between item 1 and 2 (10), should be 0/
        printTestTitle("Test: 3 content views, 1st and 2nd hidden")
        _test(views: {
            return [ContentView(), ContentView(), ContentView()]
        }, update: { stack, views in
            stack.setArrangedView(views[0], hidden: true)
            stack.setArrangedView(views[1], hidden: true)
        })
        */
    }
    
    func testRemovingItems() {
        printTestTitle("Test: 2 content views, 1st removed")
        _test(views: {
            return [ContentView(), ContentView()]
        }, update: { stack, views in
            stack.removeArrangedSubview(views[1])
        })
    }
    
    func testTextBasedViews() {
        printTestTitle("Test: 2 text based views (empty)")
        _test {
            return [UILabel(), UILabel()]
        }
        
        printTestTitle("Test: 2 text based views (one not empty)")
        _test {
            let label1 = UILabel()
            label1.text = "Test"
            let label2 = UILabel()
            label2.text = "Test"
            return [label1, label2, UILabel()]
        }
    }
    
    // MARK: Tests Implementation

    func _test(_ views: @escaping (() -> [UIView])) {
        _test(views: views, update: nil)
    }
    
    func _test(views: @escaping (() -> [UIView]), update: ((StackViewAdapter, [UIView]) -> Void)?) {
        var failedCount = 0
        let combinations = StackTestConfiguraton.generate()
        combinations.forEach {
            if !_test(views, update: update, conf: $0) {
                failedCount += 1
            }
        }
        print("Current pass: \(combinations.count - failedCount)/\(combinations.count) combinations")
        
        testCasesCount += combinations.count
        failedTestCasesCount += failedCount
        
        print("Total passes: \(testCasesCount - failedTestCasesCount)/\(testCasesCount) combinations")
    }
    
    func _test(_ viewsClosure: @escaping (() -> [UIView]), update: ((StackViewAdapter, [UIView]) -> Void)?, conf: StackTestConfiguraton) -> Bool {
        let stack1 = UIStackView()
        let stack2 = StackView()
        
        func constraints<T>(_ stack: T) -> [NSLayoutConstraint] where T: UIView, T: StackViewAdapter {
            let views = viewsClosure()
            views.enumerated().forEach {
                // This is important, tag is requried to match constraints later
                $1.tag = $0
                $1.test_isContentView = true
                $1.accessibilityIdentifier = "content-view-\($0)"
                stack.addArrangedSubview($1)
            }
            stack.axis = conf.axis
            stack.ar_alignment = conf.alignment
            stack.ar_distribution = conf.distribution
            stack.isBaselineRelativeArrangement = conf.isBaselineRelativeArrangement
            stack.isLayoutMarginsRelativeArrangement = conf.isLayoutMarginsRelativeArrangement
            stack.spacing = conf.spacing
            
            if let update = update {
                update(stack, views)
            }
            
            stack.translatesAutoresizingMaskIntoConstraints = false
            stack.updateConstraints()
            return constraintsFor(stack)
        }
        
        XCTAssertEqual(stack1.axis, stack2.axis)
        XCTAssertEqual(stack1.axis, stack2.axis)
        XCTAssertEqual(stack1.ar_alignment, stack2.ar_alignment)
        XCTAssertEqual(stack1.ar_distribution, stack2.ar_distribution)
        XCTAssertEqual(stack1.isLayoutMarginsRelativeArrangement, stack2.isLayoutMarginsRelativeArrangement)
        XCTAssertEqual(stack1.isBaselineRelativeArrangement, stack2.isBaselineRelativeArrangement)
        XCTAssertEqual(stack1.spacing, stack2.spacing)
        XCTAssertEqual(stack1.arrangedSubviews.count, stack2.arrangedSubviews.count)
        
        let constraints1 = constraints(stack1)
        let constraints2 = constraints(stack2)
        let success = assertEqualConstraints(constraints1, constraints2)
        if !success {
            print("\nFailed configuration:\n\n\(conf)")
            
            print("UIStackView constraints (count: \(constraints1.count)):\n")
            _print(constraints1)
            
            print("Arranged.StackView constraints (count: \(constraints2.count)):\n")
            _print(constraints2)
            
            print("\n===============================================")
        }
        return success
    }
    
    // MARK: Helpers
    
    func printTestTitle(_ string: String) {
        print("\n\n===============================================")
        print(string)
    }
    
    func _print(_ constraints: [NSLayoutConstraint]) {
        constraintsPrinted += 1
        guard constraintsPrinted < maxConstraintsPrinted else {
            return
        }
        print("Constraints (count: \(constraints.count)):")
        constraints.forEach {
            print($0)
        }
    }
}

class ContentView: UIView {
    var contentSize: CGSize = CGSize(width: 44, height: 44)
    convenience init(contentSize: CGSize) {
        self.init(frame: CGRect.zero)
        self.contentSize = contentSize
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override var intrinsicContentSize: CGSize {
        return self.contentSize
    }
}


Sources/Internal.swift
// The MIT License (MIT)
//
// Copyright (c) 2018 Alexander Grebenyuk (github.com/kean).

import UIKit


class LayoutGuide: UIView {
    override init(frame: CGRect) {
        super.init(frame: frame)
        // Make sure that layout guides don't interfere with touches
        self.isHidden = true
        self.translatesAutoresizingMaskIntoConstraints = false
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override class var layerClass: Swift.AnyClass {
        return CATransformLayer.self
    }
    
    override var backgroundColor: UIColor? {
        get { return nil }
        set { return }
    }
}

class GapLayoutGuide: LayoutGuide {}

class LayoutSpacer: LayoutGuide {}


extension Sequence {
    // FIXME: Name might be misleading, it doesn't enumerate over all combinations
    func forPair(_ closure: (_ first: Self.Iterator.Element, _ second: Self.Iterator.Element) -> Void) {
        let _ = reduce(nil as Self.Iterator.Element?) { previous, current in
            if let previous = previous {
                closure(previous, current)
            }
            return current
        }
    }
}


extension NSLayoutAttribute {
    var toMargin: NSLayoutAttribute {
        switch self {
        case .left: return .leftMargin
        case .right: return .rightMargin
        case .top: return .topMargin
        case .bottom: return .bottomMargin
        case .leading: return .leadingMargin
        case .trailing: return .trailingMargin
        case .centerX: return .centerXWithinMargins
        case .centerY: return .centerYWithinMargins
        default: return self
        }
    }
}


Sources/LayoutArrangement.swift
// The MIT License (MIT)
//
// Copyright (c) 2018 Alexander Grebenyuk (github.com/kean).

import UIKit

class LayoutArrangement {
    weak var canvas: UIView!
    var items = [UIView]() // Arranged views
    var hiddenItems = Set<UIView>()
    var visibleItems = Array<UIView>()
    
    var axis: UILayoutConstraintAxis = .horizontal
    var marginsEnabled: Bool = false

    private var constraints = [NSLayoutConstraint]()

    init(canvas: StackView) {
        self.canvas = canvas
    }

    func updateConstraints() {
        visibleItems = items.filter { !isHidden($0) }
        NSLayoutConstraint.deactivate(constraints.filter { $0.isActive })
        constraints.removeAll()
    }
    
    @discardableResult func constraint(item item1: UIView,
        attribute attr1: NSLayoutAttribute,
        toItem item2: UIView? = nil,
        attribute attr2: NSLayoutAttribute? = nil,
        relation: NSLayoutRelation = .equal,
        multiplier: CGFloat = 1,
        constant c: CGFloat = 0,
        priority: UILayoutPriority? = nil,
        identifier: String) -> NSLayoutConstraint
    {
        let constraint = NSLayoutConstraint(item: item1, attribute: attr1, relatedBy: relation, toItem: item2, attribute: (attr2 != nil ? attr2! : attr1), multiplier: multiplier, constant: c)
        if let priority = priority {
            constraint.priority = priority
        }
        constraint.identifier = identifier
        constraint.isActive = true
        constraints.append(constraint)
        return constraint
    }
    
    // MARK: Helpers
    
    func isHidden(_ item: UIView) -> Bool {
        return hiddenItems.contains(item)
    }
    
    func addCanvasFitConstraint(attribute: NSLayoutAttribute) {
        constraint(item: canvas, attribute: attribute, constant: 0, priority: UILayoutPriority(rawValue: 49), identifier: "ASV-canvas-fit")
    }
    
    func connectToCanvas(_ item: UIView, attribute attr: NSLayoutAttribute, weak: Bool = false) {
        let relation = connectionRelation(attr, weak: weak)
        constraint(item: canvas, attribute: (marginsEnabled ? attr.toMargin : attr), toItem: item, attribute: attr, relation: relation, identifier: "ASV-canvas-connection")
    }
    
    func connectionRelation(_ attr: NSLayoutAttribute, weak: Bool) -> NSLayoutRelation {
        if !weak { return .equal }
        return (attr == .top || attr == .left || attr == .leading) ? .lessThanOrEqual : .greaterThanOrEqual
    }
}


Sources/AlignedLayoutArrangement.swift
// The MIT License (MIT)
//
// Copyright (c) 2018 Alexander Grebenyuk (github.com/kean).

import UIKit

/** Manages alignment: constraints perpendicular to the axis.
 */
class AlignedLayoutArrangement: LayoutArrangement {
    var type: StackViewAlignment = .fill
    private var spacer: LayoutSpacer
    
    override init(canvas: StackView) {
        spacer = LayoutSpacer()
        spacer.accessibilityIdentifier = "ASV-alignment-spanner"
        
        super.init(canvas: canvas)
    }
    
    override func updateConstraints() {
        super.updateConstraints()
        
        spacer.removeFromSuperview()

        if items.count > 0 {
            updateCanvasConnectingConstraints()
            updateAlignmentConstraints()
        }
        if isAnyItemConnectionWeak {
            addItemsAmbiguitySuppressors(items)
        }
        if isAnyCanvasConnectionWeak {
            addCanvasFitConstraint(attribute: height)
        }
    }
    
    private func updateCanvasConnectingConstraints() {
        if shouldCreateSpacer {
            canvas.addSubview(spacer)
            if isAnyItemConnectionWeak {
                constraint(item: spacer, attribute: height, constant: 0, priority: UILayoutPriority(rawValue: 51), identifier: "ASV-spanning-fit")
            }
            connectItemsToSpacer(spacer, items: visibleItems, topWeak: isTopItemConnectionWeak, bottomWeak: isBottomItemConnectionWeak)
        }
        
        // FIXME: Make more readable
        let firstItem = visibleItems.count == 0 ? spacer : items.first!
        var topItem = firstItem
        var bottomItem = shouldCreateSpacer ? spacer : firstItem
        if typeIn([.center, .trailing, .lastBaseline]) {
            swap(&topItem, &bottomItem)
        }
        
        connectToCanvas(topItem, attribute: top, weak: isTopCanvasConnectionWeak)
        connectToCanvas(bottomItem, attribute: bottom, weak: isBottomCanvasConnectionWeak)
        
        if type == .center {
            connectToCanvas(firstItem, attribute: center)
        }
    }
    
    private var isAnyCanvasConnectionWeak: Bool {
        return isTopCanvasConnectionWeak || isBottomCanvasConnectionWeak
    }
    
    private var isTopCanvasConnectionWeak: Bool {
        if shouldCreateSpacer {
            return type == .firstBaseline && visibleItems.count > 0 && axis == .horizontal // .FirstBaseline specific
        }
        return isTopItemConnectionWeak
    }
    
    private var isBottomCanvasConnectionWeak: Bool {
        if shouldCreateSpacer {
            return type == .lastBaseline && visibleItems.count > 0 && axis == .horizontal // .LastBaseline specific
        }
        return isBottomItemConnectionWeak
    }
    
    private var shouldCreateSpacer: Bool {
        return visibleItems.count == 0 || (items.count > 1 && isAnyItemConnectionWeak)
    }
    
    private var isAnyItemConnectionWeak: Bool {
        return isTopItemConnectionWeak || isBottomItemConnectionWeak
    }
    
    private var isTopItemConnectionWeak: Bool {
        return typeIn([.trailing, .center, .firstBaseline, .lastBaseline])
    }
    
    private var isBottomItemConnectionWeak: Bool {
        return typeIn([.leading, .center, .firstBaseline, .lastBaseline])
    }
    
    private func updateAlignmentConstraints() {
        func attributes() -> [NSLayoutAttribute] {
            switch type {
            case .fill: return [bottom, top]
            case .leading: return [top]
            case .trailing: return [bottom]
            case .center: return [center]
            case .firstBaseline: return axis == .horizontal ? [.firstBaseline] : []
            case .lastBaseline: return axis == .horizontal ? [.lastBaseline] : []
            }
        }
        attributes().forEach {
            alignItems(items, attribute: $0)
        }
    }

    // MARK: Managed Attributes

    private var height: NSLayoutAttribute {
        return axis == .horizontal ? .height : .width
    }

    private var top: NSLayoutAttribute {
        return axis == .horizontal ? .top : .leading
    }

    private var bottom: NSLayoutAttribute {
        return axis == .horizontal ? .bottom : .trailing
    }

    private var center: NSLayoutAttribute {
        return axis == .horizontal ? .centerY : .centerX
    }
    
    // MARK: Helpers
    
    private func alignItems(_ items: [UIView], attribute: NSLayoutAttribute) {
        let firstItem = items.first!
        items.dropFirst().forEach {
            constraint(item: firstItem, attribute: attribute, toItem: $0, attribute: nil, identifier: "ASV-alignment")
        }
    }
    
    private func connectItemsToSpacer(_ spacer: LayoutSpacer, items: [UIView], topWeak: Bool, bottomWeak: Bool) {
        func connectToSpacer(_ item: UIView, attribute attr: NSLayoutAttribute, weak: Bool) {
            let relation = connectionRelation(attr, weak: weak)
            let priority: UILayoutPriority? = weak ? nil : UILayoutPriority(rawValue: 999.5)
            constraint(item: spacer, attribute: attr, toItem: item, relation: relation, priority: priority, identifier: "ASV-spanning-boundary")
        }
        items.forEach {
            connectToSpacer($0, attribute: top, weak: topWeak)
            connectToSpacer($0, attribute: bottom, weak: bottomWeak)
        }
    }
    
    private func addItemsAmbiguitySuppressors(_ items: [UIView]) {
        items.forEach {
            constraint(item: $0, attribute: height, constant: 0, priority: UILayoutPriority(rawValue: 25), identifier: "ASV-ambiguity-suppression")
        }
    }
    
    private func typeIn(_ types: [StackViewAlignment]) -> Bool {
        return types.contains(type)
    }
}
